### ISSUE TITLE
logging-format-interpolation for f-strings

### ISSUE THREAD

**Issue by eddie-dunn** [PyCQA/pylint#1788]
Pylint will warn against using .format and % when logging. As the example below shows, I think it would be prudent to also output warnings when using f-strings.

### Steps to reproduce
```python
"""Docstring"""
import logging


def main():
    """Main"""
    url = 'http://example.com/%fe'
    # This works:
    logging.error(
        'url is %s | log method: %s', url, 'logging with parameters only')

    # This will crash the logger, but pylint will output
    # logging-format-interpolation:
    logging.error(
        'url is {} | log method: %s'.format(url), 'logging with .format')

    # This will also crash the logger, but pylint accepts it:
    logging.error(f'url is {url} | log method: %s', 'logging with f-string')


main()
```

### Current behavior
Pylint does not warn against f-string interpolation in loggers.

### Expected behavior
Pylint should warn against f-string interpolation in loggers, just as it warns against .format-interpolation.

### pylint --version output
```
pylint 1.8.1, 
astroid 1.6.0
Python 3.6.2 (default, Sep 22 2017, 12:45:49) 
[GCC 6.3.0 20170516]
```

**Comment by PCManticore** [comment]
Thanks!

**Comment by sscherfke** [comment]
Please not!  f-Strings are [more readable and fast](https://twitter.com/hynek/status/939077060475211778). We don't want to mix old-style and new-style formatting in our code (this is IMHO worse than not using `logger.info('hello %s', name)`.

**Comment by sscherfke** [comment]
@PCManticore Or give it at least a different error message (e.g., `logging-fstring-interpolation`) so that users can keep using f-Strings.

**Comment by PCManticore** [comment]
A different message sounds good to me, if you can please create an issue with it, that would be great! Nevertheless, f-strings might be fast, but there's no need to compute them if there is no need to emit the logging message if its under the expected logging level.

**Comment by eddie-dunn** [comment]
@sscherfke as @PCManticore hinted at, the reason that pylint emits the logging-format-interpolation warning is because it is inefficient to compute the string for a log message that won't be logged. This also applies to f-strings, since even if they are faster than `.format` it is still slower to interpolate than not interpolate at all.

I share your opinion on the greatness of f-strings, but if you are using them with the logging module, you are using the logging module incorrectly/inefficiently.

**Comment by sscherfke** [comment]
Sometimes I just don care.  Practicality beats purity. 😉

**Comment by zachliu** [comment]
I agree with @sscherfke. Code readability out weights the tiny efficiency increase.

**Comment by leahein** [comment]
Seconded.  IMO mixing old-style formatting and f-strings throughout the application is less desirable than the cost of computing the string interpolation.

**Comment by PCManticore** [comment]
Commenting on this issue won't change anything, if you don't agree with the warning, feel free to disable it, but this is going to be the default in pylint.

**Comment by Kilo59** [comment]
@PCManticore I believe there is a bug with disabling this particular warning in the latest version of Pylint.
I'll be opening an issue shortly with additional details.

**Comment by pzelnip** [comment]
Please see the timing results at https://github.com/PyCQA/pylint/issues/2395   Those findings indicate that the difference in performance when logging messages are suppressed is minimal at best (~1%), and if logging statements *are* emitted then f-strings are actually more performant (though again, the performance difference is so small that it's not worth preferring either for performance reasons).

**Comment by joshburkart** [comment]
Yeah, this warning seems like hardcore premature optimization.

> Commenting on this issue won't change anything, if you don't agree with the warning, feel free to disable it, but this is going to be the default in pylint.

These kinds of decisions for defaults should be made by consensus rather than by fiat. Also, the imperious tone is not appreciated.

**Comment by FrankDMartinez** [comment]
How about an option to exempt f-strings? By default, they would be flagged; with this option on, no.

**Comment by cra** [comment]
You know what's even faster? Not logging at all and not writing any messages.

How about an option on pylint prudent behaviour for any stdout output?

**Comment by uriva** [comment]
The benefits of using f-strings in logs are proven - enhanced readability and less writing effort.

The downsides are highly speculative; only come into play if logging does not really occur && the f-string contains costly operations.

Controversial coding style practices should not be addressed in such a common dev tool.
("Controversial" being a huge understatement here to judge by the comments and votes)

primum non nocere

**Comment by chdsbd** [comment]
It seems like everyone here is talking about performance. There are other benefits to letting `logging` do the interpolation.

1. if interpolation fails, it will be logged instead of raising an exception
2. things like Sentry can aggregate log messages. If you interpolate before logging, Sentry would see it as one event [0]

[0]: https://reinout.vanrees.org/weblog/2015/06/05/logging-formatting.html#comment-2063573645

**Comment by sscherfke** [comment]
This all has nothing to do with pylint.  

If each kind of interpolation raises a different pylint warning, users can customize pylint to match their world-view on logging.  The discussion which one is the best can be conducted elsewhere. ;-)

**Comment by kootenpv** [comment]
@sscherfke I think the discussion is ultimately what should be the default for pylint.

**Comment by afreepenguin** [comment]
Is there a way I can customize my own pylint config to ignore this particular warning?

**Comment by Pierre-Sassoulas** [comment]
The configuration for fstring is avaialble since pylint 2.4 and was introduced in https://github.com/PyCQA/pylint/commit/eeca6cf0baa7263ff46598f06e4eda0f3072262b.

See, http://pylint.pycqa.org/en/latest/whatsnew/2.4.html?highlight=fstring%20interpolation

>  Allow the option of f-strings as a valid logging string formatting method.
> 
> logging-fstring--interpolation has been merged into logging-format-interpolation to allow the logging-format-style option to control which logging string format style is valid. To allow this, a new fstr value is valid for the logging-format-style option.

I'm going to lock to prevent further discussions here.
